### DISCUSSION TITLE
Text extraction incomplete on Linux (resolution: missing system font)

### DISCUSSION THREAD

**Discussion by maivan-hoa** [pypdfium2-team/pypdfium2#288] [Q&A] [answered]
The library's text extraction produces different results when used on Win 11 and Ubuntu. I use the same version pypdfium2==4.22.0
Here is the code that I use:

```python
import pypdfium2

docs = []
path_file = '...'
pdf_reader = pypdfium2.PdfDocument(path_file, autoclose=True)

for page_number, page in enumerate(pdf_reader, start=1):
    text_page = page.get_textpage()
    content = text_page.get_text_range()
    docs.append(content)

    text_page.close()
    page.close()

pdf_reader.close()
```

- Result on Win 11
```
Tín hiệu chuông được đánh giá là đạt khi có tín hiệu chuông ở hai hướng.
```

- Result on Ubuntu
```
Tín hiệu chuông đ ợc đánh giá là đạt khi có tín hiệu chuông ở hai h ớng.
```

How can I fix this error?
Thanks so much for taking a look!

**Comment by mara004** [comment]
Our `get_text_range()` wrapper just forwards the output from a [pdfium API function](https://github.com/pypdfium2-team/pypdfium2/blob/a1a1d378d1a5ced1c972fb6b29586bb718b11056/src/pypdfium2/_helpers/textpage.py#L91) and decodes using UTF-16LE[^1] as indicated by pdfium docs.

Unfortunately I can't really comment on the behavior of external code.
Yet, here are some questions that might help debug the issue:
* Can you share the document and page index in question?
* Am I right to assume the Windows output is correct but the Ubuntu output wrong (missing letter `ư`) ?
* Can you verify that the raw (binary) output of `FPDFText_GetText()` also differs, to rule out a decoding or shell/viewer issue?
* Can you check whether `get_text_bounded()` behaves correctly?
* Can you update to the latest version (`4.25.0` / `6219` AOTW) and retry?
* Which version of Ubuntu are you using?

[^1]: That is, python's implementation of UTF-16LE

**Reply by maivan-hoa** [reply]
Thank you for your response!
- This is the pdf page where I got the error [page_error.pdf](https://github.com/pypdfium2-team/pypdfium2/files/13850187/page_error.pdf)
- Yes, the extracted text in some sentences is missing the letter 'ư' or 'đ', ... and sometimes words are separated by spaces when running the code on Ubuntu. On Win 11 I don't have that problem
- I updated to version 4.25.0, the error still occurs
- I also tried `get_text_bounded()`, that error still occurs
- I use ubuntu which has the following version:
```
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy
```

**Reply by mara004** [reply]
Thanks for sharing the PDF.
I can reproduce the issue with `pypdfium2 extract-text` on Fedora 37.
However, I cannot reproduce it when copying the text out of Chromium (which is backed by pdfium) on the same system -- which is odd. Does Chromium make any font-related config calls that we don't?

**Reply by mara004** [reply]
What worsens the issue, pypdfium2 not only extracts text incorrectly, it also misrenders the file completely (see [page_error_1.png](https://github.com/pypdfium2-team/pypdfium2/assets/65915611/e214f453-558c-456b-a1e8-cb3f17f9e8d6)). Same observation with Okular/poppler. Chromium renders it correctly, however.

I found [pdfium_font_linux.cc](https://source.chromium.org/chromium/chromium/src/+/main:pdf/pdfium/pdfium_font_linux.cc) and the `FPDF*` calls therein, but this seems to be highly sophisticated and woven into chromium, and therefore out of my understanding.
If there's anything we have to do on the embedder side, I'd need some hint – or ideally a patch.

**Reply by maivan-hoa** [reply]
Have you tried it on windows operating system?

**Reply by mara004** [reply]
That said, I wonder if this issue could possibly be handled by the end user installing some system font package?

**Reply by maivan-hoa** [reply]
Yes so do I. I will try installing some more fonts into the system to test. Thank you!

**Reply by mara004** [reply]
> Have you tried it on windows operating system?

No, I don't have access to windows (except CI)  -- but according to your report it works, right?
Maybe Windows ships fonts Ubuntu doesn't by default, or maybe pdfium font search behaves differently on Linux. I don't know.

**Reply by mara004** [reply, marked as answer]
I believe the culprit might be proprietary Windows fonts such as Arial and TimesNewRoman which distros can't ship for licensing reasons - so the PDF viewer substitutes with some other available font, which might not have the special chars.
Presumably it would work if you copy over the Windows fonts to Ubuntu.

See the attached inspection [screenshot](https://github.com/pypdfium2-team/pypdfium2/assets/65915611/87e94863-97c6-4c10-8132-fe7b81bb5fcf) from Okular

**Reply by maivan-hoa** [reply]
I installed the font into my Ubuntu system and it worked. Thank you for your help!
